modules/packages/Sort.chpl

```diff
@@ -1319,12 +1570,1106 @@ module RadixSortHelp {
       }
     }
   }
+
+  record RadixBucketizer {
+    proc getNumBuckets() {
+      return (1 << RADIX_BITS) + 2; // +2 for end-before and end-after bins
+    }
+    proc getBinsToRecursivelySort() {
+      // first and last bin represent end-of-key and do not need
+      // to be recursively sorted.
+      return 1..(1 << RADIX_BITS);
+    }
+    proc getNextStartBit(start_bit:int) {
+      return start_bit+8;
+    }
+
+    // returns the bin for classifying a single record
+    inline proc bucketForRecord(a, criterion, startbit) {
+      const (bin, _) = binForRecord(a, criterion, startbit);
+      return bin;
+    }
+
+    // yields (index, bucket index) for A[start_n..end_n]
+    iter classify(A, start_n, end_n, criterion, startbit) {
+      var cur = start_n;
+      while cur <= end_n-(classifyUnrollFactor-1) {
+        for /*param*/ j in 0..classifyUnrollFactor-1 {
+          yield (cur+j, bucketForRecord(A[cur+j], criterion, startbit));
+        }
+        cur += classifyUnrollFactor;
+      }
+      while cur <= end_n {
+        yield (cur, bucketForRecord(A[cur], criterion, startbit));
+        cur += 1;
+      }
+    }
+  }
+}
+
+pragma "no doc"
+module ShallowCopy {
+
+  // These shallow copy functions "move" a record around
+  // (i.e. they neither swap nor call a copy initializer).
+
+  // TODO: move these out of the Sort module and/or consider
+  // language support for it.
+
+  // TODO: These shallowCopy functions should handle Block,Cyclic arrays
+  inline proc shallowCopy(ref A, dst, src, nElts) {
+    // Ideally this would just be
+    //A[dst..#nElts] = A[src..#nElts];
+
+    if boundsChecking {
+      assert(nElts > 0);
+      assert(A.domain.contains(dst..#nElts));
+      assert(A.domain.contains(src..#nElts));
+    }
+
+    if A._instance.isDefaultRectangular() {
+      var size = (nElts:size_t)*c_sizeof(A.eltType);
+      c_memcpy(c_ptrTo(A[dst]), c_ptrTo(A[src]), size);
```

Comment:
If there were a possibility of overlap, this should use `memmove` instead of `memcpy`.  But AFAICT this `shallowCopy` is presently used only with `size=1`.

Similarly below, if `DstA` and `SrcA` might be the same array.  (Or slices or views of the same array.)